# Capstone Project - The Battle of Neighbourhood
# Neighbourhood recommendation system for immigrants planning to move to Toronto

### Installing and Importing python libraries

In [2]:
!pip install geocoder
!pip install folium

     |████████████████████████████████| 98 kB 9.1 MB/s  eta 0:00:01
     |████████████████████████████████| 93 kB 4.6 MB/s  eta 0:00:01


In [3]:
import pandas as pd
import requests
import numpy as np
import geocoder
import folium
import requests 
import matplotlib.cm as cm
import matplotlib.colors as colors
import json
import xml
import matplotlib.pyplot as plt
%matplotlib inline
import warnings
warnings.filterwarnings("ignore")

from pandas.io.json import json_normalize 
from sklearn.cluster import KMeans
from geopy.geocoders import Nominatim 
from bs4 import BeautifulSoup

pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', None)

### Data Wrangling

Importing data from wikipedia page to a Data Frame.

Link: https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M

In [9]:
data_page = requests.get('https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M')
df = pd.read_html(data_page.content, header=0)[0]

In [10]:
df = df[df['Borough'] != 'Not assigned']

In [13]:
df['Neighbourhood'] = np.where(df['Neighbourhood'] == 'Not assigned', df['Borough'], df['Neighbourhood'])
df.head()

Postal Code           Borough                                Neighbourhood
2         M3A        North York                                    Parkwoods
3         M4A        North York                             Victoria Village
4         M5A  Downtown Toronto                    Regent Park, Harbourfront
5         M6A        North York             Lawrence Manor, Lawrence Heights
6         M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government

In [14]:
df.shape

(103, 3)

Import Latitude and Longitude values from file

In [15]:
coordinates_df = pd.read_csv('http://cocl.us/Geospatial_data')
coordinates_df.head()

Postal Code   Latitude  Longitude
0         M1B  43.806686 -79.194353
1         M1C  43.784535 -79.160497
2         M1E  43.763573 -79.188711
3         M1G  43.770992 -79.216917
4         M1H  43.773136 -79.239476

In [18]:
df_toronto = pd.merge(df, coordinates_df, on = 'Postal Code')
df_toronto.head()

Postal Code           Borough                                Neighbourhood  \
0         M3A        North York                                    Parkwoods   
1         M4A        North York                             Victoria Village   
2         M5A  Downtown Toronto                    Regent Park, Harbourfront   
3         M6A        North York             Lawrence Manor, Lawrence Heights   
4         M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government   

    Latitude  Longitude  
0  43.753259 -79.329656  
1  43.725882 -79.315572  
2  43.654260 -79.360636  
3  43.718518 -79.464763  
4  43.662301 -79.389494

Toronto coordinates were used to set the center location:

In [19]:
address = 'Toronto, TO'

geolocator = Nominatim(user_agent="TO_explorer")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The geograpical coordinate of Toronto are {}, {}.'.format(latitude, longitude))

The geograpical coordinate of Toronto are 43.65238435, -79.38356765.


Mapping the neighbourhoods based on their coordinates (latitude and longitude).

In [22]:
map_toronto_neighbourhoods = folium.Map(location=[latitude, longitude], zoom_start=10)


for lat, lng, label in zip(df_toronto['Latitude'], df_toronto['Longitude'], df_toronto['Neighbourhood']):
    label = '{}'.format(label)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_toronto_neighbourhoods)  
    
map_toronto_neighbourhoods

Define Foursquare credentials and version

In [23]:
CLIENT_ID = '5O3XNCTGA1HG5H442L15ZOYHLBBPYDFT5ZUZYYAKIOC5NHVH' # your Foursquare ID
CLIENT_SECRET = 'ADMQYLOHUDOD5DRUOK1VIKNZAD5QI0HKAM0S3USEEYKUEVCD' # your Foursquare Secret
VERSION = '20190326' # Foursquare API version
LIMIT = 100 # A default Foursquare API limit value

print('Your credentails:')
print('CLIENT_ID: ' + CLIENT_ID)
print('CLIENT_SECRET:' + CLIENT_SECRET)

Your credentails:
CLIENT_ID: 5O3XNCTGA1HG5H442L15ZOYHLBBPYDFT5ZUZYYAKIOC5NHVH
CLIENT_SECRET:ADMQYLOHUDOD5DRUOK1VIKNZAD5QI0HKAM0S3USEEYKUEVCD


Create a function which will return near by venues

In [24]:
def getNearbyVenues(names, latitudes, longitudes, radius=500,LIMIT=500):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
            
       
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            LIMIT)
            
      
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighbourhood', 
                  'Neighbourhood Latitude', 
                  'Neighbourhood Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    
    return(nearby_venues)

The function is run for all neighbourhoods in Toronto and a new dataframe called Toronto_venues is created

In [25]:
Toronto_venues = getNearbyVenues(names=df_toronto['Neighbourhood'],
                                   latitudes=df_toronto['Latitude'],
                                   longitudes=df_toronto['Longitude']
                                 )

Parkwoods
Victoria Village
Regent Park, Harbourfront
Lawrence Manor, Lawrence Heights
Queen's Park, Ontario Provincial Government
Islington Avenue, Humber Valley Village
Malvern, Rouge
Don Mills
Parkview Hill, Woodbine Gardens
Garden District, Ryerson
Glencairn
West Deane Park, Princess Gardens, Martin Grove, Islington, Cloverdale
Rouge Hill, Port Union, Highland Creek
Don Mills
Woodbine Heights
St. James Town
Humewood-Cedarvale
Eringate, Bloordale Gardens, Old Burnhamthorpe, Markland Wood
Guildwood, Morningside, West Hill
The Beaches
Berczy Park
Caledonia-Fairbanks
Woburn
Leaside
Central Bay Street
Christie
Cedarbrae
Hillcrest Village
Bathurst Manor, Wilson Heights, Downsview North
Thorncliffe Park
Richmond, Adelaide, King
Dufferin, Dovercourt Village
Scarborough Village
Fairview, Henry Farm, Oriole
Northwood Park, York University
East Toronto, Broadview North (Old East York)
Harbourfront East, Union Station, Toronto Islands
Little Portugal, Trinity
Kennedy Park, Ionview, East Birchmo

Number of venues created for each neighbourhood

In [26]:
Toronto_venues.groupby('Neighbourhood').count()

Neighbourhood Latitude  \
Neighbourhood                                                                
Agincourt                                                                4   
Alderwood, Long Branch                                                   7   
Bathurst Manor, Wilson Heights, Downsview North                         21   
Bayview Village                                                          4   
Bedford Park, Lawrence Manor East                                       23   
Berczy Park                                                             57   
Birch Cliff, Cliffside West                                              4   
Brockton, Parkdale Village, Exhibition Place                            25   
Business reply mail Processing Centre, South Ce...                      16   
CN Tower, King and Spadina, Railway Lands, Harb...                      17   
Caledonia-Fairbanks                                                      4   
Canada Post Gateway Processing Centre                                   13   
Cedarbrae                                                                9   
Central Bay Street                                                      62   
Christie                                                                16   
Church and Wellesley                                                    79   
Clarks Corners, Tam O'Shanter, Sullivan                                 14   
Cliffside, Cliffcrest, Scarborough Village West                          2   
Commerce Court, Victoria Hotel                                         100   
Davisville                                                              36   
Davisville North                                                         8   
Del Ray, Mount Dennis, Keelsdale and Silverthorn                         5   
Don Mills                                                               23   
Dorset Park, Wexford Heights, Scarborough Town ...                       7   
Downsview                                                               14   
Dufferin, Dovercourt Village                                            15   
East Toronto, Broadview North (Old East York)                            3   
Eringate, Bloordale Gardens, Old Burnhamthorpe,...                       9   
Fairview, Henry Farm, Oriole                                            68   
First Canadian Place, Underground city                                 100   
Forest Hill North & West, Forest Hill Road Park                          4   
Garden District, Ryerson                                               100   
Glencairn                                                                5   
Golden Mile, Clairlea, Oakridge                                         10   
Guildwood, Morningside, West Hill                                        8   
Harbourfront East, Union Station, Toronto Islands                      100   
High Park, The Junction South                                           24   
Hillcrest Village                                                        5   
Humber Summit                                                            2   
Humberlea, Emery                                                         1   
Humewood-Cedarvale                                                       3   
India Bazaar, The Beaches West                                          21   
Kennedy Park, Ionview, East Birchmount Park                              5   
Kensington Market, Chinatown, Grange Park                               62   
Kingsview Village, St. Phillips, Martin Grove G...                       4   
Lawrence Manor, Lawrence Heights                                        14   
Lawrence Park                                                            4   
Leaside                                                                 31   
Little Portugal, Trinity                                                44   
Malvern, Rouge                                                           2

Number of unique categories found from the generated data.

In [27]:
len(Toronto_venues['Venue Category'].unique())

273

### Neighborhood venues analysis

Onehot coding is used to create a dataframe with occurence of each venue present (1) or not (0) under each Neighborhood

In [36]:
Toronto_onehot = pd.get_dummies(Toronto_venues[['Venue Category']],prefix="",prefix_sep="")

Toronto_onehot['Neighbourhood'] = Toronto_venues['Neighbourhood'] 


fixed_columns = [Toronto_onehot.columns[-1]] + list(Toronto_onehot.columns[:-1])
Toronto_onehot = Toronto_onehot[fixed_columns]

Toronto_onehot.head()

Neighbourhood  Accessories Store  Adult Boutique  Afghan Restaurant  \
0         Parkwoods                  0               0                  0   
1         Parkwoods                  0               0                  0   
2  Victoria Village                  0               0                  0   
3  Victoria Village                  0               0                  0   
4  Victoria Village                  0               0                  0   

   Airport  Airport Food Court  Airport Lounge  Airport Service  \
0        0                   0               0                0   
1        0                   0               0                0   
2        0                   0               0                0   
3        0                   0               0                0   
4        0                   0               0                0   

   Airport Terminal  American Restaurant  Antique Shop  Aquarium  Art Gallery  \
0                 0                    0             0         0            0   
1                 0                    0             0         0            0   
2                 0                    0             0         0            0   
3                 0                    0             0         0            0   
4                 0                    0             0         0            0   

   Art Museum  Arts & Crafts Store  Asian Restaurant  Athletics & Sports  \
0           0                    0                 0                   0   
1           0                    0                 0                   0   
2           0                    0                 0                   0   
3           0                    0                 0                   0   
4           0                    0                 0                   0   

   Auto Garage  Auto Workshop  BBQ Joint  Baby Store  Bagel Shop  Bakery  \
0            0              0          0           0           0       0   
1            0              0          0           0           0       0   
2            0              0          0           0           0       0   
3            0              0          0           0           0       0   
4            0              0          0           0           0       0   

   Bank  Bar  Baseball Field  Baseball Stadium  Basketball Court  \
0     0    0               0                 0                 0   
1     0    0               0                 0                 0   
2     0    0               0                 0                 0   
3     0    0               0                 0                 0   
4     0    0               0                 0                 0   

   Basketball Stadium  Beach  Bed & Breakfast  Beer Bar  Beer Store  \
0                   0      0                0         0           0   
1                   0      0                0         0           0   
2                   0      0                0         0           0   
3                   0      0                0         0           0   
4                   0      0                0         0           0   

   Belgian Restaurant  Bike Shop  Bistro  Boat or Ferry  Bookstore  Boutique  \
0                   0          0       0              0          0         0   
1                   0          0       0              0          0         0   
2                   0          0       0              0          0         0   
3                   0          0       0              0          0         0   
4                   0          0       0              0          0         0   

   Brazilian Restaurant  Breakfast Spot  Brewery  Bridal Shop  \
0                     0               0        0            0   
1                     0               0        0            0   
2                     0               0        0            0   
3                     0               0        0            0   
4                     0               0        0            0   

   Bubble Tea Shop  Building  B

The rows are grouped by neighborhood and the mean of the frequency of occurrence of each category is calculated.

In [38]:
Toronto_grouped = Toronto_onehot.groupby('Neighbourhood').mean().reset_index()
Toronto_grouped.head()

Neighbourhood  Accessories Store  \
0                                        Agincourt                0.0   
1                           Alderwood, Long Branch                0.0   
2  Bathurst Manor, Wilson Heights, Downsview North                0.0   
3                                  Bayview Village                0.0   
4                Bedford Park, Lawrence Manor East                0.0   

   Adult Boutique  Afghan Restaurant  Airport  Airport Food Court  \
0             0.0                0.0      0.0                 0.0   
1             0.0                0.0      0.0                 0.0   
2             0.0                0.0      0.0                 0.0   
3             0.0                0.0      0.0                 0.0   
4             0.0                0.0      0.0                 0.0   

   Airport Lounge  Airport Service  Airport Terminal  American Restaurant  \
0             0.0              0.0               0.0             0.000000   
1             0.0              0.0               0.0             0.000000   
2             0.0              0.0               0.0             0.000000   
3             0.0              0.0               0.0             0.000000   
4             0.0              0.0               0.0             0.043478   

   Antique Shop  Aquarium  Art Gallery  Art Museum  Arts & Crafts Store  \
0           0.0       0.0          0.0         0.0                  0.0   
1           0.0       0.0          0.0         0.0                  0.0   
2           0.0       0.0          0.0         0.0                  0.0   
3           0.0       0.0          0.0         0.0                  0.0   
4           0.0       0.0          0.0         0.0                  0.0   

   Asian Restaurant  Athletics & Sports  Auto Garage  Auto Workshop  \
0               0.0                 0.0          0.0            0.0   
1               0.0                 0.0          0.0            0.0   
2               0.0                 0.0          0.0            0.0   
3               0.0                 0.0          0.0            0.0   
4               0.0                 0.0          0.0            0.0   

   BBQ Joint  Baby Store  Bagel Shop  Bakery      Bank  Bar  Baseball Field  \
0        0.0         0.0         0.0     0.0  0.000000  0.0             0.0   
1        0.0         0.0         0.0     0.0  0.000000  0.0             0.0   
2        0.0         0.0         0.0     0.0  0.095238  0.0             0.0   
3        0.0         0.0         0.0     0.0  0.250000  0.0             0.0   
4        0.0         0.0         0.0     0.0  0.000000  0.0             0.0   

   Baseball Stadium  Basketball Court  Basketball Stadium  Beach  \
0               0.0               0.0                 0.0    0.0   
1               0.0               0.0                 0.0    0.0   
2               0.0               0.0                 0.0    0.0   
3               0.0               0.0                 0.0    0.0   
4               0.0               0.0                 0.0    0.0   

   Bed & Breakfast  Beer Bar  Beer Store  Belgian Restaurant  Bike Shop  \
0              0.0       0.0         0.0                 0.0        0.0   
1              0.0       0.0         0.0                 0.0        0.0   
2              0.0       0.0         0.0                 0.0        0.0   
3              0.0       0.0         0.0                 0.0        0.0   
4              0.0       0.0         0.0                 0.0        0.0   

   Bistro  Boat or Ferry  Bookstore  Boutique  Brazilian Restaurant  \
0     0.0            0.0        0.0  0.000000                   0.0   
1     0.0            0.0        0.0  0.000000                   0.0   
2     0.0            0.0        0.0  0.000000                   0.0   
3     0.0            0.0        0.0  0.000000                   0.0   
4     0.0            0.0        0.0  0.043478                   0.0   

   Breakfast Spot  Brewery  Bridal Shop  Bubble Tea Shop  Building  \
0         

Set Neighbourhood as index.

In [41]:
Toronto_grouped.set_index('Neighbourhood', inplace=True)
Toronto_grouped.head()

Accessories Store  \
Neighbourhood                                                        
Agincourt                                                      0.0   
Alderwood, Long Branch                                         0.0   
Bathurst Manor, Wilson Heights, Downsview North                0.0   
Bayview Village                                                0.0   
Bedford Park, Lawrence Manor East                              0.0   

                                                 Adult Boutique  \
Neighbourhood                                                     
Agincourt                                                   0.0   
Alderwood, Long Branch                                      0.0   
Bathurst Manor, Wilson Heights, Downsview North             0.0   
Bayview Village                                             0.0   
Bedford Park, Lawrence Manor East                           0.0   

                                                 Afghan Restaurant  Airport  \
Neighbourhood                                                                 
Agincourt                                                      0.0      0.0   
Alderwood, Long Branch                                         0.0      0.0   
Bathurst Manor, Wilson Heights, Downsview North                0.0      0.0   
Bayview Village                                                0.0      0.0   
Bedford Park, Lawrence Manor East                              0.0      0.0   

                                                 Airport Food Court  \
Neighbourhood                                                         
Agincourt                                                       0.0   
Alderwood, Long Branch                                          0.0   
Bathurst Manor, Wilson Heights, Downsview North                 0.0   
Bayview Village                                                 0.0   
Bedford Park, Lawrence Manor East                               0.0   

                                                 Airport Lounge  \
Neighbourhood                                                     
Agincourt                                                   0.0   
Alderwood, Long Branch                                      0.0   
Bathurst Manor, Wilson Heights, Downsview North             0.0   
Bayview Village                                             0.0   
Bedford Park, Lawrence Manor East                           0.0   

                                                 Airport Service  \
Neighbourhood                                                      
Agincourt                                                    0.0   
Alderwood, Long Branch                                       0.0   
Bathurst Manor, Wilson Heights, Downsview North              0.0   
Bayview Village                                              0.0   
Bedford Park, Lawrence Manor East                            0.0   

                                                 Airport Terminal  \
Neighbourhood                                                       
Agincourt                                                     0.0   
Alderwood, Long Branch                                        0.0   
Bathurst Manor, Wilson Heights, Downsview North               0.0   
Bayview Village                                               0.0   
Bedford Park, Lawrence Manor East                             0.0   

                                                 American Restaurant  \
Neighbourhood                                                          
Agincourt                                                   0.000000   
Alderwood, Long Branch                                      0.000000   
Bathurst Manor, Wilson Heights, Downsview North             0.000000   
Bayview Village                                             0.000000   
Bedford Park, Lawrence Manor East                           0.043478   

                                                 Antique Shop  Aquarium  \
Neighbourhood            

In [42]:
Toronto_grouped.shape

(94, 272)

### Feature Selection
Currently there are 272 venue categories in the dataset with a lot of them belonging to the same general "group". i.e there is "Vietnamese Restaurant", "Greek Restaurant", "Vegetarian Restaurant" all belonging to the same logical group. The 272 venues will be filtered through a key words search so to create a number of higher level categories which will be used for both clustering as well as recommendation.

For this purpose the categories created with corresponding filters are:

1. Restaurants - keywords (Restaurant|Steak|Bistro)
2. Fast_Food - keywords (Fast|Burger|Food Truck|Sandwich|Pizza Place|Place|Joint)
3. Health_Fitness - keywords (Yoga|Gym|Martial|Health|Fitness|Pool|Healthy|Field|Sports|Trail|Stadium|Tennis|Climbing)
4. Kids_Friendly - keywords (Kid|Play|Park|Nursery|Playground|School|College|Nursery|Medical Center|Clinic|Basketball|Football)
5. Nightlife' - keywords (Night|Bar|Pub|Wine|Dance|Strip Club|Adult|Club|Speakeasy|Liquor|Brewery)
6. Culture - keywords (Gallery|Museum|Paintings|Monument|Church|Theater|Cinema|Sculpture|Aquarium|Concert|Art)
7. Shopping - keywords (Mall|Shopping|Shop|Retail|Store|Boutique)
8. Short_stay - keywords (Motel|Hotel|Airport|Stay)
9. Personal_care - keywords (Salon|Barber|Spa|Massage|Hair|Nail|Tanning)
10. Food_markets - keywords (Supermarket|Market|Grocery|Bakery|Butcher)
11. Transportation - keywords (Train|Metro|Bus|Tram|Boat)
12. Leisure - keywords (Beach|Plaza|Skating)

In [43]:
pd.options.mode.chained_assignment = None

Toronto_1=Toronto_grouped.loc[:, Toronto_grouped.columns.str.contains('Restaurant|Steak|Bistro')]
Toronto_1['total1']=Toronto_1.sum(axis=1)

Toronto_2=Toronto_grouped.loc[:, Toronto_grouped.columns.str.contains('Fast|Burger|Food Truck|Sandwich|Pizza Place|Place|Joint')]
Toronto_2['total2']=Toronto_2.sum(axis=1)

Toronto_3=Toronto_grouped.loc[:, Toronto_grouped.columns.str.contains('Yoga|Gym|Martial|Health|Fitness|Pool|Healthy|Field|Sports|Trail|Stadium|Tennis|Climbing')]
Toronto_3['total3']=Toronto_3.sum(axis=1)

Toronto_4=Toronto_grouped.loc[:, Toronto_grouped.columns.str.contains('Kid|Play|Park|Nursery|Playground|School|College|Nursery|Medical Center|Clinic|Basketball|Football')]
Toronto_4['total4']=Toronto_4.sum(axis=1)

Toronto_5=Toronto_grouped.loc[:, Toronto_grouped.columns.str.contains('Night|Bar|Pub|Wine|Dance|Strip Club|Adult|Club|Speakeasy|Liquor|Brewery')]
Toronto_5['total5']=Toronto_5.sum(axis=1)

Toronto_6=Toronto_grouped.loc[:, Toronto_grouped.columns.str.contains('Gallery|Museum|Paintings|Monument|Church|Theater|Cinema|Sculpture|Aquarium|Concert|Art')]
Toronto_6['total6']=Toronto_6.sum(axis=1)

Toronto_7=Toronto_grouped.loc[:, Toronto_grouped.columns.str.contains('Mall|Shopping|Shop|Retail|Store|Boutique')]
Toronto_7['total7']=Toronto_7.sum(axis=1)

Toronto_8=Toronto_grouped.loc[:, Toronto_grouped.columns.str.contains('Motel|Hotel|Airport|Stay')]
Toronto_8['total8']=Toronto_8.sum(axis=1)

Toronto_9=Toronto_grouped.loc[:, Toronto_grouped.columns.str.contains('Salon|Barber|Spa|Massage|Hair|Nail|Tanning')]
Toronto_9['total9']=Toronto_9.sum(axis=1)

Toronto_10=Toronto_grouped.loc[:, Toronto_grouped.columns.str.contains('Supermarket|Market|Grocery|Bakery|Butcher')]
Toronto_10['total10']=Toronto_10.sum(axis=1)

Toronto_11=Toronto_grouped.loc[:, Toronto_grouped.columns.str.contains('Train|Metro|Bus|Tram|Boat')]
Toronto_11['total11']=Toronto_11.sum(axis=1)

Toronto_12=Toronto_grouped.loc[:, Toronto_grouped.columns.str.contains('Beach|Plaza|Skating')]
Toronto_12['total12']=Toronto_12.sum(axis=1)

In [44]:
Toronto_grouped['Restaurants']=Toronto_1['total1']
Toronto_grouped['Fast_Food']=Toronto_2['total2']
Toronto_grouped['Health_Fitness']=Toronto_3['total3']
Toronto_grouped['Kids_Friendly']=Toronto_4['total4']
Toronto_grouped['Nightlife']=Toronto_5['total5']
Toronto_grouped['Culture']=Toronto_6['total6']
Toronto_grouped['Shopping']=Toronto_7['total7']
Toronto_grouped['Short_stay']=Toronto_8['total8']
Toronto_grouped['Personal_care']=Toronto_9['total9']
Toronto_grouped['Food_markets']=Toronto_10['total10']
Toronto_grouped['Transportation']=Toronto_11['total11']
Toronto_grouped['Leisure']=Toronto_12['total12']
Toronto_grouped.head(5)

Accessories Store  \
Neighbourhood                                                        
Agincourt                                                      0.0   
Alderwood, Long Branch                                         0.0   
Bathurst Manor, Wilson Heights, Downsview North                0.0   
Bayview Village                                                0.0   
Bedford Park, Lawrence Manor East                              0.0   

                                                 Adult Boutique  \
Neighbourhood                                                     
Agincourt                                                   0.0   
Alderwood, Long Branch                                      0.0   
Bathurst Manor, Wilson Heights, Downsview North             0.0   
Bayview Village                                             0.0   
Bedford Park, Lawrence Manor East                           0.0   

                                                 Afghan Restaurant  Airport  \
Neighbourhood                                                                 
Agincourt                                                      0.0      0.0   
Alderwood, Long Branch                                         0.0      0.0   
Bathurst Manor, Wilson Heights, Downsview North                0.0      0.0   
Bayview Village                                                0.0      0.0   
Bedford Park, Lawrence Manor East                              0.0      0.0   

                                                 Airport Food Court  \
Neighbourhood                                                         
Agincourt                                                       0.0   
Alderwood, Long Branch                                          0.0   
Bathurst Manor, Wilson Heights, Downsview North                 0.0   
Bayview Village                                                 0.0   
Bedford Park, Lawrence Manor East                               0.0   

                                                 Airport Lounge  \
Neighbourhood                                                     
Agincourt                                                   0.0   
Alderwood, Long Branch                                      0.0   
Bathurst Manor, Wilson Heights, Downsview North             0.0   
Bayview Village                                             0.0   
Bedford Park, Lawrence Manor East                           0.0   

                                                 Airport Service  \
Neighbourhood                                                      
Agincourt                                                    0.0   
Alderwood, Long Branch                                       0.0   
Bathurst Manor, Wilson Heights, Downsview North              0.0   
Bayview Village                                              0.0   
Bedford Park, Lawrence Manor East                            0.0   

                                                 Airport Terminal  \
Neighbourhood                                                       
Agincourt                                                     0.0   
Alderwood, Long Branch                                        0.0   
Bathurst Manor, Wilson Heights, Downsview North               0.0   
Bayview Village                                               0.0   
Bedford Park, Lawrence Manor East                             0.0   

                                                 American Restaurant  \
Neighbourhood                                                          
Agincourt                                                   0.000000   
Alderwood, Long Branch                                      0.000000   
Bathurst Manor, Wilson Heights, Downsview North             0.000000   
Bayview Village                                             0.000000   
Bedford Park, Lawrence Manor East                           0.043478   

                                                 Antique Shop  Aquarium  \
Neighbourhood            

In [45]:
Toronto_grouped_new= Toronto_grouped[['Restaurants',
                                    'Fast_Food',
                                    'Health_Fitness',
                                    'Kids_Friendly',
                                    'Nightlife',
                                    'Culture',
                                    'Shopping',
                                    'Short_stay',
                                    'Personal_care',
                                    'Food_markets',
                                    'Transportation',
                                    'Leisure']].copy()
Toronto_grouped_new.head()

Restaurants  Fast_Food  \
Neighbourhood                                                             
Agincourt                                           0.250000   0.000000   
Alderwood, Long Branch                              0.000000   0.428571   
Bathurst Manor, Wilson Heights, Downsview North     0.142857   0.142857   
Bayview Village                                     0.500000   0.000000   
Bedford Park, Lawrence Manor East                   0.391304   0.130435   

                                                 Health_Fitness  \
Neighbourhood                                                     
Agincourt                                              0.000000   
Alderwood, Long Branch                                 0.142857   
Bathurst Manor, Wilson Heights, Downsview North        0.000000   
Bayview Village                                        0.000000   
Bedford Park, Lawrence Manor East                      0.000000   

                                                 Kids_Friendly  Nightlife  \
Neighbourhood                                                               
Agincourt                                                  0.0   0.000000   
Alderwood, Long Branch                                     0.0   0.142857   
Bathurst Manor, Wilson Heights, Downsview North            0.0   0.000000   
Bayview Village                                            0.0   0.000000   
Bedford Park, Lawrence Manor East                          0.0   0.130435   

                                                 Culture  Shopping  \
Neighbourhood                                                        
Agincourt                                            0.0  0.000000   
Alderwood, Long Branch                               0.0  0.142857   
Bathurst Manor, Wilson Heights, Downsview North      0.0  0.380952   
Bayview Village                                      0.0  0.000000   
Bedford Park, Lawrence Manor East                    0.0  0.260870   

                                                 Short_stay  Personal_care  \
Neighbourhood                                                                
Agincourt                                               0.0            0.0   
Alderwood, Long Branch                                  0.0            0.0   
Bathurst Manor, Wilson Heights, Downsview North         0.0            0.0   
Bayview Village                                         0.0            0.0   
Bedford Park, Lawrence Manor East                       0.0            0.0   

                                                 Food_markets  Transportation  \
Neighbourhood                                                                   
Agincourt                                            0.000000             0.0   
Alderwood, Long Branch                               0.000000             0.0   
Bathurst Manor, Wilson Heights, Downsview North      0.095238             0.0   
Bayview Village                                      0.000000             0.0   
Bedford Park, Lawrence Manor East                    0.086957             0.0   

                                                 Leisure  
Neighbourhood                                             
Agincourt                                           0.25  
Alderwood, Long Branch                              0.00  
Bathurst Manor, Wilson Heights, Downsview North     0.00  
Bayview Village                                     0.00  
Bedford Park, Lawrence Manor East                   0.00

In [46]:
Toronto_grouped_new.shape

(94, 12)

Let's print each neighborhood along with the top 5 most common venues
First the Neighborhood is returned as a column in the dataframe

In [47]:
Toronto_grouped_new.reset_index(level=0,inplace=True)

In [52]:
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]


num_top_venues = 5

indicators = ['st', 'nd', 'rd']


columns = ['Neighbourhood']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))


neighbourhoods_venues_sorted = pd.DataFrame(columns=columns)
neighbourhoods_venues_sorted['Neighbourhood'] = Toronto_grouped_new['Neighbourhood']

for ind in np.arange(Toronto_grouped_new.shape[0]):
    neighbourhoods_venues_sorted.iloc[ind, 1:] = return_most_common_venues(Toronto_grouped_new.iloc[ind, :], num_top_venues)

neighbourhoods_venues_sorted.head(10)


Neighbourhood 1st Most Common Venue  \
0                                          Agincourt               Leisure   
1                             Alderwood, Long Branch             Fast_Food   
2    Bathurst Manor, Wilson Heights, Downsview North              Shopping   
3                                    Bayview Village           Restaurants   
4                  Bedford Park, Lawrence Manor East           Restaurants   
5                                        Berczy Park              Shopping   
6                        Birch Cliff, Cliffside West               Leisure   
7       Brockton, Parkdale Village, Exhibition Place              Shopping   
8  Business reply mail Processing Centre, South C...             Fast_Food   
9  CN Tower, King and Spadina, Railway Lands, Har...            Short_stay   

  2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
0           Restaurants        Transportation          Food_markets   
1              Shopping             Nightlife        Health_Fitness   
2             Fast_Food           Restaurants          Food_markets   
3               Leisure        Transportation          Food_markets   
4              Shopping             Nightlife             Fast_Food   
5           Restaurants             Nightlife          Food_markets   
6         Kids_Friendly        Health_Fitness        Transportation   
7        Health_Fitness          Food_markets               Culture   
8          Food_markets         Kids_Friendly           Restaurants   
9              Shopping        Transportation               Culture   

  5th Most Common Venue  
0         Personal_care  
1               Leisure  
2               Leisure  
3         Personal_care  
4          Food_markets  
5               Culture  
6          Food_markets  
7             Nightlife  
8         Personal_care  
9             Nightlife

In [55]:
neighbourhoods_venues_sorted.to_csv("top5.csv", index=False)

### K-means Clustering

Preprocessing the data.

In [56]:
from sklearn.preprocessing import StandardScaler
X = Toronto_grouped_new.values[:,1:].astype(float)
X = np.nan_to_num(X)
cluster_dataset = StandardScaler().fit_transform(X)
cluster_dataset

array([[ 0.39487793, -0.74915845, -0.52822975, ..., -0.63863091,
        -0.24607752,  5.32995767],
       [-1.05411213,  2.60901663,  0.40111789, ..., -0.63863091,
        -0.24607752, -0.26010574],
       [-0.22611781,  0.37023325, -0.52822975, ...,  0.54511916,
        -0.24607752, -0.26010574],
       ...,
       [ 0.87787462, -0.74915845, -0.52822975, ..., -0.63863091,
        -0.24607752, -0.26010574],
       [-1.05411213, -0.74915845,  0.40111789, ..., -0.63863091,
         1.64822465,  2.93421621],
       [-1.05411213, -0.74915845, -0.52822975, ..., -0.63863091,
        -0.24607752, -0.26010574]])

In [57]:
from sklearn.cluster import KMeans
num_clusters = 5
k_means = KMeans(init="k-means++", n_clusters=num_clusters, n_init=12)
k_means.fit(cluster_dataset)
labels = k_means.labels_

print(labels)

[4 4 4 4 4 1 3 1 4 0 3 4 4 4 4 1 4 0 1 4 4 4 4 1 4 1 3 4 4 1 2 4 1 2 4 1 1
 3 3 3 3 1 4 1 2 4 2 4 1 4 1 4 3 4 3 4 1 4 3 1 4 3 4 1 1 3 4 1 4 4 3 4 1 4
 4 1 4 4 1 1 4 4 4 1 1 4 4 4 4 4 4 4 2 3]


In [58]:
Toronto_grouped_new["Labels"] = labels
Toronto_grouped_new.head()

Neighbourhood  Restaurants  Fast_Food  \
0                                        Agincourt     0.250000   0.000000   
1                           Alderwood, Long Branch     0.000000   0.428571   
2  Bathurst Manor, Wilson Heights, Downsview North     0.142857   0.142857   
3                                  Bayview Village     0.500000   0.000000   
4                Bedford Park, Lawrence Manor East     0.391304   0.130435   

   Health_Fitness  Kids_Friendly  Nightlife  Culture  Shopping  Short_stay  \
0        0.000000            0.0   0.000000      0.0  0.000000         0.0   
1        0.142857            0.0   0.142857      0.0  0.142857         0.0   
2        0.000000            0.0   0.000000      0.0  0.380952         0.0   
3        0.000000            0.0   0.000000      0.0  0.000000         0.0   
4        0.000000            0.0   0.130435      0.0  0.260870         0.0   

   Personal_care  Food_markets  Transportation  Leisure  Labels  
0            0.0      0.000000             0.0     0.25       4  
1            0.0      0.000000             0.0     0.00       4  
2            0.0      0.095238             0.0     0.00       4  
3            0.0      0.000000             0.0     0.00       4  
4            0.0      0.086957             0.0     0.00       4

Merge data frames Toronto_grouped_new and df_toronto

In [59]:
toronto_merged = Toronto_grouped_new
toronto_merged = toronto_merged.join(df_toronto.set_index('Neighbourhood'), on='Neighbourhood')
toronto_merged.head()

Neighbourhood  Restaurants  Fast_Food  \
0                                        Agincourt     0.250000   0.000000   
1                           Alderwood, Long Branch     0.000000   0.428571   
2  Bathurst Manor, Wilson Heights, Downsview North     0.142857   0.142857   
3                                  Bayview Village     0.500000   0.000000   
4                Bedford Park, Lawrence Manor East     0.391304   0.130435   

   Health_Fitness  Kids_Friendly  Nightlife  Culture  Shopping  Short_stay  \
0        0.000000            0.0   0.000000      0.0  0.000000         0.0   
1        0.142857            0.0   0.142857      0.0  0.142857         0.0   
2        0.000000            0.0   0.000000      0.0  0.380952         0.0   
3        0.000000            0.0   0.000000      0.0  0.000000         0.0   
4        0.000000            0.0   0.130435      0.0  0.260870         0.0   

   Personal_care  Food_markets  Transportation  Leisure  Labels Postal Code  \
0            0.0      0.000000             0.0     0.25       4         M1S   
1            0.0      0.000000             0.0     0.00       4         M8W   
2            0.0      0.095238             0.0     0.00       4         M3H   
3            0.0      0.000000             0.0     0.00       4         M2K   
4            0.0      0.086957             0.0     0.00       4         M5M   

       Borough   Latitude  Longitude  
0  Scarborough  43.794200 -79.262029  
1    Etobicoke  43.602414 -79.543484  
2   North York  43.754328 -79.442259  
3   North York  43.786947 -79.385975  
4   North York  43.733283 -79.419750

In [60]:
toronto_merged.reset_index(level=0, inplace=True)

In [80]:
toronto_merged.to_csv("clusters.csv", index=False)

index                                      Neighbourhood  Restaurants  \
0       0                                          Agincourt     0.250000   
1       1                             Alderwood, Long Branch     0.000000   
2       2    Bathurst Manor, Wilson Heights, Downsview North     0.142857   
3       3                                    Bayview Village     0.500000   
4       4                  Bedford Park, Lawrence Manor East     0.391304   
5       5                                        Berczy Park     0.245614   
6       6                        Birch Cliff, Cliffside West     0.000000   
7       7       Brockton, Parkdale Village, Exhibition Place     0.080000   
8       8  Business reply mail Processing Centre, South C...     0.125000   
9       9  CN Tower, King and Spadina, Railway Lands, Har...     0.000000   
10     10                                Caledonia-Fairbanks     0.000000   
11     11              Canada Post Gateway Processing Centre     0.230769   
12     12                                          Cedarbrae     0.333333   
13     13                                 Central Bay Street     0.274194   
14     14                                           Christie     0.125000   
15     15                               Church and Wellesley     0.367089   
16     16            Clarks Corners, Tam O'Shanter, Sullivan     0.357143   
17     17    Cliffside, Cliffcrest, Scarborough Village West     0.500000   
18     18                     Commerce Court, Victoria Hotel     0.320000   
19     19                                         Davisville     0.277778   
20     20                                   Davisville North     0.000000   
21     21   Del Ray, Mount Dennis, Keelsdale and Silverthorn     0.200000   
22     22                                          Don Mills     0.347826   
23     22                                          Don Mills     0.347826   
24     23  Dorset Park, Wexford Heights, Scarborough Town...     0.571429   
25     24                                          Downsview     0.000000   
26     24                                          Downsview     0.000000   
27     24                                          Downsview     0.000000   
28     24                                          Downsview     0.000000   
29     25                       Dufferin, Dovercourt Village     0.066667   
30     26      East Toronto, Broadview North (Old East York)     0.000000   
31     27  Eringate, Bloordale Gardens, Old Burnhamthorpe...     0.000000   
32     28                       Fairview, Henry Farm, Oriole     0.161765   
33     29             First Canadian Place, Underground city     0.340000   
34     30    Forest Hill North & West, Forest Hill Road Park     0.250000   
35     31                           Garden District, Ryerson     0.210000   
36     32                                          Glencairn     0.200000   
37     33                    Golden Mile, Clairlea, Oakridge     0.000000   
38     34                  Guildwood, Morningside, West Hill     0.250000   
39     35  Harbourfront East, Union Station, Toronto Islands     0.150000   
40     36                      High Park, The Junction South     0.291667   
41     37                                  Hillcrest Village     0.200000   
42     38                                      Humber Summit     0.000000   
43     39                                   Humberlea, Emery     0.000000   
44     40                                 Humewood-Cedarvale     0.000000   
45     41                     India Bazaar, The Beaches West     0.285714   
46     42        Kennedy Park, Ionview, East Birchmount Park     0.000000   
47     43          Kensington Market, Chinatown, Grange Park     0.290323   
48     44  Kingsview Village, St. Phillips, Martin Grove ...     0.000000   
49     45                   Lawrence Manor, Lawrence Heights     0.071429   
50     46                                      Lawrence Park     0.000000   


In [66]:
map_clusters = folium.Map(location=[latitude, longitude], zoom_start=11)

# set color scheme for the clusters
x = np.arange(num_clusters)
ys = [i + x + (i*x)**2 for i in range(num_clusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(toronto_merged['Latitude'], toronto_merged['Longitude'], toronto_merged['Neighbourhood'], toronto_merged['Labels']):
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[cluster-1],
        fill=True,
        fill_color=rainbow[cluster-1],
        fill_opacity=1).add_to(map_clusters)
       
map_clusters

Let's feed the system with user ratings for each of the 12 categories we created depending on how important each was for the user in a scale of 1-10. For the testing purpose we used a “persona” of a user that is married with children and rated each category accordingly.

In [68]:
userInput_Family = [
            {'Venue':'Restaurants','Rating':7},
            {'Venue':'Fast_Food','Rating':5},
            {'Venue':'Health_Fitness','Rating':8},
            {'Venue':'Kids_Friendly','Rating':10},
            {'Venue':'Nightlife','Rating':1},
            {'Venue':'Culture','Rating':8},
            {'Venue':'Shopping','Rating':9},
            {'Venue':'Short_stay','Rating':0},
            {'Venue':'Personal_care','Rating':6},
            {'Venue':'Food_markets','Rating':10},
            {'Venue':'Transportation','Rating':5},
            {'Venue':'Leisure','Rating':8}
         ] 
inputVenues_Family = pd.DataFrame(userInput_Family) 

inputVenues_Family

Venue  Rating
0      Restaurants       7
1        Fast_Food       5
2   Health_Fitness       8
3    Kids_Friendly      10
4        Nightlife       1
5          Culture       8
6         Shopping       9
7       Short_stay       0
8    Personal_care       6
9     Food_markets      10
10  Transportation       5
11         Leisure       8

In [69]:
inputVenues_Family.set_index('Venue', inplace=True)

In [71]:
Toronto_grouped_new.set_index('Neighbourhood',inplace=True)

In [72]:
Toronto_grouped_new.head()

Restaurants  Fast_Food  \
Neighbourhood                                                             
Agincourt                                           0.250000   0.000000   
Alderwood, Long Branch                              0.000000   0.428571   
Bathurst Manor, Wilson Heights, Downsview North     0.142857   0.142857   
Bayview Village                                     0.500000   0.000000   
Bedford Park, Lawrence Manor East                   0.391304   0.130435   

                                                 Health_Fitness  \
Neighbourhood                                                     
Agincourt                                              0.000000   
Alderwood, Long Branch                                 0.142857   
Bathurst Manor, Wilson Heights, Downsview North        0.000000   
Bayview Village                                        0.000000   
Bedford Park, Lawrence Manor East                      0.000000   

                                                 Kids_Friendly  Nightlife  \
Neighbourhood                                                               
Agincourt                                                  0.0   0.000000   
Alderwood, Long Branch                                     0.0   0.142857   
Bathurst Manor, Wilson Heights, Downsview North            0.0   0.000000   
Bayview Village                                            0.0   0.000000   
Bedford Park, Lawrence Manor East                          0.0   0.130435   

                                                 Culture  Shopping  \
Neighbourhood                                                        
Agincourt                                            0.0  0.000000   
Alderwood, Long Branch                               0.0  0.142857   
Bathurst Manor, Wilson Heights, Downsview North      0.0  0.380952   
Bayview Village                                      0.0  0.000000   
Bedford Park, Lawrence Manor East                    0.0  0.260870   

                                                 Short_stay  Personal_care  \
Neighbourhood                                                                
Agincourt                                               0.0            0.0   
Alderwood, Long Branch                                  0.0            0.0   
Bathurst Manor, Wilson Heights, Downsview North         0.0            0.0   
Bayview Village                                         0.0            0.0   
Bedford Park, Lawrence Manor East                       0.0            0.0   

                                                 Food_markets  Transportation  \
Neighbourhood                                                                   
Agincourt                                            0.000000             0.0   
Alderwood, Long Branch                               0.000000             0.0   
Bathurst Manor, Wilson Heights, Downsview North      0.095238             0.0   
Bayview Village                                      0.000000             0.0   
Bedford Park, Lawrence Manor East                    0.086957             0.0   

                                                 Leisure  Labels  
Neighbourhood                                                     
Agincourt                                           0.25       4  
Alderwood, Long Branch                              0.00       4  
Bathurst Manor, Wilson Heights, Downsview North     0.00       4  
Bayview Village                                     0.00       4  
Bedford Park, Lawrence Manor East                   0.00       4

Next we created the user preferences matrix by calculating the weighted average of the user ratings against the frequency of each category and for all Toronto neighbourhoods in our dataframe. The result was sorted to a top 5 list of recommended neighbourhoods basis user profile which were then plotted on the map of Toronto.

In [73]:
recommendationTable_family = ((Toronto_grouped_new*inputVenues_Family['Rating']).sum(axis=1))/(inputVenues_Family['Rating'].sum())
recommendationTable_family.head(5)

Neighbourhood
Agincourt                                          0.048701
Alderwood, Long Branch                             0.061224
Bathurst Manor, Wilson Heights, Downsview North    0.079159
Bayview Village                                    0.045455
Bedford Park, Lawrence Manor East                  0.087521
dtype: float64

We add the Neighborhood coordinates to map our results:

In [75]:
TOR= recommendationTable_family.to_frame().reset_index()
recommendT=pd.merge(df_toronto, TOR, on="Neighbourhood")
result=recommendT.head(5)
result

Postal Code           Borough                                Neighbourhood  \
0         M3A        North York                                    Parkwoods   
1         M4A        North York                             Victoria Village   
2         M5A  Downtown Toronto                    Regent Park, Harbourfront   
3         M6A        North York             Lawrence Manor, Lawrence Heights   
4         M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government   

    Latitude  Longitude         0  
0  43.753259 -79.329656  0.123377  
1  43.725882 -79.315572  0.072727  
2  43.654260 -79.360636  0.082157  
3  43.718518 -79.464763  0.111317  
4  43.662301 -79.389494  0.082386

In [77]:
neighborhoods_venues_sorted.loc[neighborhoods_venues_sorted['Neighbourhood'].isin(result['Neighbourhood'])]

Neighbourhood 1st Most Common Venue  \
45             Lawrence Manor, Lawrence Heights              Shopping   
61                                    Parkwoods              Shopping   
62  Queen's Park, Ontario Provincial Government              Shopping   
63                    Regent Park, Harbourfront              Shopping   
85                             Victoria Village           Restaurants   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
45         Personal_care        Health_Fitness           Restaurants   
61         Kids_Friendly               Leisure        Transportation   
62           Restaurants         Kids_Friendly             Fast_Food   
63             Nightlife          Food_markets               Culture   
85              Shopping             Fast_Food               Leisure   

   5th Most Common Venue  
45               Leisure  
61          Food_markets  
62             Nightlife  
63         Kids_Friendly  
85        Transportation

In [82]:
compare=neighbourhoods_venues_sorted.loc[neighbourhoods_venues_sorted['Neighbourhood'].isin(result['Neighbourhood'])]
compare

Neighbourhood 1st Most Common Venue  \
45             Lawrence Manor, Lawrence Heights              Shopping   
61                                    Parkwoods              Shopping   
62  Queen's Park, Ontario Provincial Government              Shopping   
63                    Regent Park, Harbourfront              Shopping   
85                             Victoria Village           Restaurants   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
45         Personal_care        Health_Fitness           Restaurants   
61         Kids_Friendly               Leisure        Transportation   
62           Restaurants         Kids_Friendly             Fast_Food   
63             Nightlife          Food_markets               Culture   
85              Shopping             Fast_Food               Leisure   

   5th Most Common Venue  
45               Leisure  
61          Food_markets  
62             Nightlife  
63         Kids_Friendly  
85        Transportation

In [78]:
recommendTmap=pd.merge(df_toronto, TOR, on="Neighbourhood")
recommendTmap['Labels']=toronto_merged['Labels']
result_labels=recommendTmap.head(5)
result_labels

Postal Code           Borough                                Neighbourhood  \
0         M3A        North York                                    Parkwoods   
1         M4A        North York                             Victoria Village   
2         M5A  Downtown Toronto                    Regent Park, Harbourfront   
3         M6A        North York             Lawrence Manor, Lawrence Heights   
4         M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government   

    Latitude  Longitude         0  Labels  
0  43.753259 -79.329656  0.123377       4  
1  43.725882 -79.315572  0.072727       4  
2  43.654260 -79.360636  0.082157       4  
3  43.718518 -79.464763  0.111317       4  
4  43.662301 -79.389494  0.082386       4

In [79]:
map_recommended_neighbourhoods = folium.Map(location=[latitude, longitude], zoom_start=11)


x = np.arange(num_clusters)
ys = [i + x + (i*x)**2 for i in range(num_clusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

markers_colors = []
for lat, lon, poi, cluster in zip(result_labels['Latitude'], result_labels['Longitude'], result_labels['Neighbourhood'], result_labels['Labels']):
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[cluster-1],
        fill=True,
        fill_color=rainbow[cluster-1],
        fill_opacity=1).add_to(map_recommended_neighbourhoods)
       
map_recommended_neighbourhoods